File: apps/web-v2/src/lib/view/conversation-preview-input.ts
import type { LastMessageAttachmentSummary, PreviewAttachmentKind } from '@meeshy/shared/types/conversation-preview';
import type {
  ConversationPreviewAttachment,
  ConversationPreviewInput,
  ConversationPreviewMessage,
} from '@meeshy/shared/utils/conversation-preview';

import type { ListConversation, ListLastMessage } from '@/lib/api/list-preview';
import type { Attachment, Conversation } from '@/lib/api/types';

/**
 * CE QUE LA LIGNE SAIT EN PLUS DE LA CONVERSATION — le lecteur, ses langues,
 * l'instant, et les trois sources locales que le cache de liste ne porte pas :
 * la réception d'un éphémère (#7451), la frappe, le brouillon.
 */
export type PreviewContext = {
  readonly viewerId: string;
  /** Langue de CADRAGE (celle de l'interface) : celle des libellés. */
  readonly language: string;
  readonly preferredLanguages: readonly string[];
  readonly now: number;
  readonly receivedAt?: number | null;
  /** L'échéance servie par `message:countdown-started`, quand elle est connue. */
  readonly servedDeadline?: number | null;
  readonly typing?: readonly string[] | null;
  readonly draft?: string | null;
};

type Hoisted = {
  readonly forwardedFromId?: string;
  readonly maxViewOnceCount?: number;
  readonly location?: { readonly name?: string | null; readonly address?: string | null } | null;
};

type ProtectedAttachment = Attachment & {
  readonly isViewOnce?: boolean;
  readonly isBlurred?: boolean;
  readonly effectFlags?: number;
};

const nonEmpty = (value: string | null | undefined): string | null =>
  typeof value === 'string' && value.trim() !== '' ? value : null;

const kindOfMime = (mimeType: string): PreviewAttachmentKind => {
  if (mimeType.startsWith('image/')) return 'image';
  if (mimeType.startsWith('video/')) return 'video';
  if (mimeType.startsWith('audio/')) return 'audio';
  return 'file';
};

/**
 * LE RÉSUMÉ DES PIÈCES, QUAND LE SERVEUR NE L'A PAS SERVI — un message reçu
 * par `message:new` ou mon envoi portent leurs pièces ENTIÈRES, pas le résumé
 * que `GET /conversations` joint à `lastMessage`. La même projection que le
 * contrat (#7545) : familles par `mimeType`, somme des tailles connues.
 */
function summaryOf(attachments: readonly Attachment[]): LastMessageAttachmentSummary | null {
  if (attachments.length === 0) return null;
  const kinds = attachments.reduce<Partial<Record<PreviewAttachmentKind, number>>>((acc, a) => {
    const kind = kindOfMime(a.mimeType ?? '');
    return { ...acc, [kind]: (acc[kind] ?? 0) + 1 };
  }, {});
  const sizes = attachments.map((a) => a.fileSize).filter((s): s is number => typeof s === 'number' && Number.isFinite(s));
  return { count: attachments.length, kinds, totalSize: sizes.length === 0 ? null : sizes.reduce((a, b) => a + b, 0) };
}

function attachmentOf(attachment: ProtectedAttachment): ConversationPreviewAttachment {
  return {
    mimeType: attachment.mimeType,
    originalName: nonEmpty(attachment.originalName) ?? nonEmpty(attachment.fileName),
    fileSize: attachment.fileSize ?? null,
    duration: attachment.duration ?? null,
    width: attachment.width ?? null,
    height: attachment.height ?? null,
    pageCount: attachment.pageCount ?? null,
    ...(attachment.isViewOnce === undefined ? {} : { isViewOnce: attachment.isViewOnce }),
    ...(attachment.isBlurred === undefined ? {} : { isBlurred: attachment.isBlurred }),
    ...(attachment.effectFlags === undefined ? {} : { effectFlags: attachment.effectFlags }),
  };
}

function messageOf(conversation: Conversation, context: PreviewContext): ConversationPreviewMessage | null {
  const message = conversation.lastMessage as (ListLastMessage & Hoisted) | undefined | null;
  if (message === undefined || message === null) return null;
  const sender = message.sender;
  const attachments = (message.attachments ?? []) as readonly ProtectedAttachment[];
  const first = attachments[0];
  const summary = message.attachmentSummary ?? summaryOf(attachments);
  const served = context.servedDeadline ?? null;
  const expiresAt = served ?? message.expiresAt ?? null;
  const max = message.maxViewOnceCount;

  return {
    id: message.id,
    senderId: message.senderId ?? null,
    senderUserId: sender?.userId ?? null,
    senderName: nonEmpty(sender?.displayName) ?? nonEmpty(sender?.user?.displayName) ?? nonEmpty(sender?.user?.username),
    content: message.content ?? null,
    originalLanguage: conversation.lastMessageOriginalLanguage ?? message.originalLanguage ?? null,
    translations: conversation.lastMessageTranslations ?? null,
    createdAt: message.createdAt,
    messageType: message.messageType ?? null,
    effectFlags: message.effectFlags ?? null,
    ephemeralDuration: message.ephemeralDuration ?? null,
    expiresAt,
    isEncrypted: message.isEncrypted ?? null,
    isViewOnce: message.isViewOnce ?? null,
    isBlurred: message.isBlurred ?? null,
    viewOnceConsumed: typeof max === 'number' && max > 0 ? (message.viewOnceCount ?? 0) >= max : null,
    isForwarded: message.isForwarded ?? (typeof message.forwardedFromId === 'string' && message.forwardedFromId !== ''),
    systemEvent: message.systemEvent ?? null,
    callSummary: message.callSummary ?? null,
    location: message.location ?? null,
    attachment: first === undefined ? null : attachmentOf(first),
    attachmentSummary: summary,
  };
}

/**
 * `previewInputOf` — la conversation du cache de liste projetée sur l'ENTRÉE
 * de `composeConversationPreview` (#7546). Une PROJECTION, jamais une
 * composition : aucun libellé, aucune priorité, aucune règle de protection
 * n'est tranchée ici — tout cela appartient au composeur partagé, que le SDK
 * iOS rejoue sur le même fichier de cas.
 */
export function previewInputOf(conversation: Conversation, context: PreviewContext): ConversationPreviewInput {
  const row = conversation as ListConversation;
  return {
    viewerId: context.viewerId,
    language: context.language,
    preferredLanguages: context.preferredLanguages,
    now: context.now,
    receivedAt: context.receivedAt ?? null,
    activeCall: row.activeCall ?? null,
    typing: context.typing ?? null,
    draft: context.draft ?? null,
    lastReaction: row.lastReaction ?? null,
    lastMessage: messageOf(conversation, context),
  };
}

File: packages/MeeshySDK/Sources/MeeshySDK/Models/ConversationPreview/ConversationPreviewInput.swift
import Foundation

// L'ENTRÉE du composeur de la ligne d'aperçu — miroir de
// `ConversationPreviewInput` (`packages/shared/utils/conversation-preview.ts`).
// Les noms de champs sont ceux du TS, clé pour clé : c'est ce qui permet au
// fichier de cas commun (`packages/shared/fixtures/conversation-preview-cases.json`)
// d'être décodé tel quel par le test du SDK.

/// La première pièce jointe, avec ses détails TELS QU'ILS EXISTENT — chacun
/// n'apparaît dans la ligne que s'il est renseigné. `duration` est en
/// MILLISECONDES, comme `MessageAttachment.duration`.
public struct ConversationPreviewAttachment: Sendable, Hashable, Decodable {
    public let mimeType: String?
    public let originalName: String?
    public let fileSize: Double?
    public let duration: Double?
    public let width: Double?
    public let height: Double?
    public let pageCount: Double?
    public let isViewOnce: Bool?
    public let isBlurred: Bool?
    public let effectFlags: Int?
    /// Texte alternatif (`MessageAttachment.alt`). Pour un sticker de texte,
    /// c'est la phrase tapée, dessinée dans l'image (décision porteur 2026-09-23).
    public let alt: String?

    public init(mimeType: String? = nil, originalName: String? = nil, fileSize: Double? = nil,
                duration: Double? = nil, width: Double? = nil, height: Double? = nil,
                pageCount: Double? = nil, isViewOnce: Bool? = nil, isBlurred: Bool? = nil,
                effectFlags: Int? = nil, alt: String? = nil) {
        self.mimeType = mimeType
        self.originalName = originalName
        self.fileSize = fileSize
        self.duration = duration
        self.width = width
        self.height = height
        self.pageCount = pageCount
        self.isViewOnce = isViewOnce
        self.isBlurred = isBlurred
        self.effectFlags = effectFlags
        self.alt = alt
    }
}

/// Le sticker hissé de `metadata.sticker` : sa PRÉSENCE fait du message un
/// sticker, quel que soit son `messageType`.
public struct ConversationPreviewSticker: Sendable, Hashable, Decodable {
    public let templateId: String?
    public let emoji: String?

    public init(templateId: String? = nil, emoji: String? = nil) {
        self.templateId = templateId
        self.emoji = emoji
    }
}

/// Un lieu partagé, tel que la ligne le nomme.
public struct ConversationPreviewPlace: Sendable, Hashable, Decodable {
    public let name: String?
    public let address: String?

    public init(name: String? = nil, address: String? = nil) {
        self.name = name
        self.address = address
    }
}

/// Le dernier message tel que le client le tient — projection de
/// `lastMessage` (REST) ou des clés plates de `conversation:updated`.
public struct ConversationPreviewMessage: Sendable, Hashable, Decodable {
    public let id: String
    /// `Participant.id` ou `User.id` de l'expéditeur — l'un des deux suffit à
    /// reconnaître « Vous ».
    public let senderId: String?
    public let senderUserId: String?
    public let senderName: String?
    public let content: String?
    public let originalLanguage: String?
    public let translations: [String: String]?
    public let createdAt: Date
    public let messageType: String?
    public let effectFlags: Int?
    public let ephemeralDuration: Int?
    /// L'échéance SERVIE à ce lecteur (#7451).
    public let expiresAt: Date?
    public let isEncrypted: Bool?
    public let isViewOnce: Bool?
    public let isBlurred: Bool?
    /// Le lecteur a déjà ouvert ce message à vue unique.
    public let viewOnceConsumed: Bool?
    public let isForwarded: Bool?
    public let systemEvent: LastMessageSystemEvent?
    public let callSummary: LastMessageCallSummary?
    public let sticker: ConversationPreviewSticker?
    public let location: ConversationPreviewPlace?
    public let attachment: ConversationPreviewAttachment?
    public let attachmentSummary: LastMessageAttachmentSummary?

    public init(
        id: String,
        senderId: String? = nil,
        senderUserId: String? = nil,
        senderName: String? = nil,
        content: String? = nil,
        originalLanguage: String? = nil,
        translations: [String: String]? = nil,
        createdAt: Date,
        messageType: String? = nil,
        effectFlags: Int? = nil,
        ephemeralDuration: Int? = nil,
        expiresAt: Date? = nil,
        isEncrypted: Bool? = nil,
        isViewOnce: Bool? = nil,
        isBlurred: Bool? = nil,
        viewOnceConsumed: Bool? = nil,
        isForwarded: Bool? = nil,
        systemEvent: LastMessageSystemEvent? = nil,
        callSummary: LastMessageCallSummary? = nil,
        sticker: ConversationPreviewSticker? = nil,
        location: ConversationPreviewPlace? = nil,
        attachment: ConversationPreviewAttachment? = nil,
        attachmentSummary: LastMessageAttachmentSummary? = nil
    ) {
        self.id = id
        self.senderId = senderId
        self.senderUserId = senderUserId
        self.senderName = senderName
        self.content = content
        self.originalLanguage = originalLanguage
        self.translations = translations
        self.createdAt = createdAt
        self.messageType = messageType
        self.effectFlags = effectFlags
        self.ephemeralDuration = ephemeralDuration
        self.expiresAt = expiresAt
        self.isEncrypted = isEncrypted
        self.isViewOnce = isViewOnce
        self.isBlurred = isBlurred
        self.viewOnceConsumed = viewOnceConsumed
        self.isForwarded = isForwarded
        self.systemEvent = systemEvent
        self.callSummary = callSummary
        self.sticker = sticker
        self.location = location
        self.attachment = attachment
        self.attachmentSummary = attachmentSummary
    }
}

public struct ConversationPreviewInput: Sendable, Hashable, Decodable {
    /// `User.id` du lecteur.
    public let viewerId: String
    /// Langue de CADRAGE : celle des libellés.
    public let language: String
    /// Le prisme du lecteur, dans l'ordre (`ReaderPrism`).
    public let preferredLanguages: [String]
    public let now: Date
    /// Première réception locale du dernier message — le départ du décompte
    /// d'un éphémère (#7451).
    public let receivedAt: Date?
    public let activeCall: ConversationActiveCall?
    /// Noms des personnes qui écrivent, lecteur exclu.
    public let typing: [String]?
    public let draft: String?
    public let lastReaction: ConversationLastReaction?
    public let lastMessage: ConversationPreviewMessage?

    public init(
        viewerId: String,
        language: String,
        preferredLanguages: [String],
        now: Date,
        receivedAt: Date? = nil,
        activeCall: ConversationActiveCall? = nil,
        typing: [String]? = nil,
        draft: String? = nil,
        lastReaction: ConversationLastReaction? = nil,
        lastMessage: ConversationPreviewMessage? = nil
    ) {
        self.viewerId = viewerId
        self.language = language
        self.preferredLanguages = preferredLanguages
        self.now = now
        self.receivedAt = receivedAt
        self.activeCall = activeCall
        self.typing = typing
        self.draft = draft
        self.lastReaction = lastReaction
        self.lastMessage = lastMessage
    }
}
